arch/xtensa/core/mmu.c

```diff
@@ -0,0 +1,278 @@
+/* Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <stdint.h>
+#include <stdbool.h>
+#include <zephyr/kernel.h>
+#include <xtensa/config/core-isa.h>
+
+#define ASID_INVALID 0
+
+#define TLB_PTES_WAY    7 /* where to pin the page table mapping */
+#define TLB_VECBASE_WAY 8 /* where to pin the vecbase mapping */
+
+/* expensive check, for debugging only */
+/* #define VALIDATE_PTE_ATTRS */
+
+struct tlb_regs {
+	uint32_t rasid;
+	uint32_t ptevaddr;
+	uint32_t ptepin_as;
+	uint32_t ptepin_at;
+	uint32_t vecpin_as;
+	uint32_t vecpin_at;
+};
+
+/* Word index (not byte offset) into L1 page table */
+static uint32_t l1_idx(uint32_t addr)
+{
+	return addr >> 22;
+}
+
+/* Virtual address of the PTE mapping the specified address */
+static uint32_t pte_virt(uint32_t addr, struct tlb_regs *regs)
+{
+	return 4 * (addr >> 12) + regs->ptevaddr;
+}
+
+static bool pte_valid(uint32_t pte)
+{
+	return ((pte & 0xf) < 12);
+}
+
+static uint32_t pte_ring(uint32_t pte)
+{
+	return (pte >> 4) & 3;
+}
+
+static uint32_t pte_addr(uint32_t pte)
+{
+	return pte & ~0xfff;
+}
+
+static uint32_t lookup_pte(uint32_t *l1, uint32_t addr)
+{
+	uint32_t pte1 = l1[l1_idx(addr)];
+	uint32_t *l2 = (void *) (pte1 & ~0xfff);
+
+	return pte_valid(pte1) ? l2[(addr >> 12) & 0x3ff] : 0xf;
+}
+
+/* Walks a page table, ensuring that:
+ *
+ * 1. L1 page table entries (entries used in hardware refill) are
+ *    mapped at ring 0 and either invalid or read-only
+ *
+ * 2. Those hardware addresses mapped by page table pages are also
+ *    direct-mapped at their hardware address with ring 0 permissions.
+ *
+ * 3. The cacheability attribute of the two mappings must be
+ *    identical, and in multiprocessor environments they must be
+ *    uncached.
+ */
+void xtensa_page_table_validate(uint32_t *l1)
+{
+	for (int i = 0; i < 1024; i++) {
+		uint32_t pte = l1[i];
+
+		if (!pte_valid(pte)) {
+			continue;
+		}
+
+		__ASSERT_NO_MSG(pte_ring(pte) == 0);
+
+		uint32_t phys = lookup_pte(l1, pte_addr(pte));
+
+		__ASSERT_NO_MSG(pte_valid(phys));
+		__ASSERT_NO_MSG((phys >> 12) == (pte >> 12)); /* map to same page */
+		__ASSERT_NO_MSG(pte_ring(phys) == 0);
+		__ASSERT_NO_MSG((pte & 0xc) == (phys & 0xc)); /* cache bits identical */
+		if (CONFIG_MP_MAX_NUM_CPUS > 1) {
+			__ASSERT_NO_MSG((pte & 0xf) == 0);
+		}
+	}
+}
+
+static void compute_regs(uint32_t user_asid, uint32_t *l1_page, struct tlb_regs *regs)
+{
+	uint32_t vecbase = XTENSA_RSR("VECBASE");
+
+	__ASSERT_NO_MSG((((uint32_t) l1_page) & 0xfff) == 0);
+	__ASSERT_NO_MSG(user_asid != 1 && user_asid < CONFIG_XTENSA_MMU_ASID_MAX);
+
+	/* We don't use ring 1/2, ring 0 ASID must be 1 */
+	regs->rasid = (user_asid << 24) | 0x000001;
+
+	/* Derive PTEVADDR from ASID so each domain gets its own PTE area */
+	regs->ptevaddr = CONFIG_XTENSA_MMU_PTE_BASE + user_asid * 0x400000;
+
+	/* The ptables code doesn't add the mapping for the l1 page itself */
+	l1_page[l1_idx(regs->ptevaddr)] = (uint32_t) l1_page;
```

Comment:
Manipulating the page table here does not look nice if the goal is to split the mmu / page table logic

Reply:
the cache bits may not match as well ...